File: android/app/src/main/java/com/blainemiller/scripturealone/data/topics/LifeThemes.kt
package com.blainemiller.scripturealone.data.topics

import com.blainemiller.scripturealone.data.VerseRange
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.text.Normalizer

/** A heading in the Topics directory — "Worry & Hard Feelings", "Who God Is". */
data class LifeThemeGroup(val id: String, val name: String, val localizedName: String = name)

/**
 * Something a reader may be going through, and passages that speak to it — `LifeThemes.swift`. The
 * same curated list as iOS (`Data/topics/`, built by `Tools/build_topics.py` into `LifeThemes.json`,
 * which the build copies in), so a theme holds the same passages on both platforms.
 *
 * Passages are references only, in KJV keys: the words always come from the translation being read.
 * The English fields are the source; the localized ones come from `life_themes.xml` in each `res/values` folder.
 */
data class LifeTheme(
    /** Stable: never renamed. */
    val id: String,
    val group: String,
    val name: String,
    val description: String,
    val synonyms: List<String>,
    val passages: List<VerseRange>,
    /** Nave's Topical Bible topics that cover the same ground, by name. */
    val naveTopics: List<String>,
    val localizedName: String = name,
    val localizedDescription: String = description,
    val localizedSynonyms: List<String> = emptyList(),
)

/** The curated life themes. */
class LifeThemeCatalog(val groups: List<LifeThemeGroup>, val themes: List<LifeTheme>) {

    fun theme(id: String): LifeTheme? = themes.firstOrNull { it.id == id }
    fun themes(group: LifeThemeGroup): List<LifeTheme> = themes.filter { it.group == group.id }

    /**
     * The catalog in the app's language: [groupNames], [names], [descriptions] and [synonyms] are the
     * string arrays of `life_themes.xml`, in `LifeThemes.json` order. A list of the wrong length (a
     * stale generated file) leaves the English rather than pairing a theme with another's words.
     */
    fun localized(groupNames: List<String>, names: List<String>, descriptions: List<String>, synonyms: List<String>): LifeThemeCatalog {
        val groupsOk = groupNames.size == groups.size
        val themesOk = names.size == themes.size && descriptions.size == themes.size && synonyms.size == themes.size
        return LifeThemeCatalog(
            groups.mapIndexed { i, g -> if (groupsOk) g.copy(localizedName = groupNames[i]) else g },
            themes.mapIndexed { i, t ->
                if (!themesOk) {
                    t
                } else {
                    t.copy(
                        localizedName = names[i],
                        localizedDescription = descriptions[i],
                        localizedSynonyms = splitSynonyms(synonyms[i]),
                    )
                }
            },
        )
    }

    /**
     * Themes a search speaks to, best first — `LifeThemeCatalog.search` on iOS: the name and search
     * words, in English and the app's language, scored by [TopicSearch.score].
     */
    fun search(query: String, limit: Int = 3): List<LifeTheme> {
        val needle = TopicSearch.normalize(query)
        if (needle.length < TopicSearch.minimumLength(needle)) return emptyList()
        return themes.withIndex().mapNotNull { (index, theme) ->
            val terms = listOf(theme.name, theme.localizedName) + theme.synonyms + theme.localizedSynonyms
            val score = terms.maxOf { TopicSearch.score(TopicSearch.normalize(it), needle) }
            if (score > 0) Triple(score, index, theme) else null
        }.sortedWith(compareByDescending<Triple<Int, Int, LifeTheme>> { it.first }.thenBy { it.second })
            .take(limit)
            .map { it.third }
    }

    companion object {
        val EMPTY = LifeThemeCatalog(emptyList(), emptyList())

        /** `LifeThemes.json`: groups, then themes with their passages as storage strings. */
        fun parse(json: String): LifeThemeCatalog {
            val root = Json.parseToJsonElement(json).jsonObject
            fun JsonObject.string(key: String) = get(key)?.jsonPrimitive?.content.orEmpty()
            fun JsonObject.strings(key: String) = (get(key) as? JsonArray)?.map { it.jsonPrimitive.content }.orEmpty()
            val groups = root.getValue("groups").jsonArray.map { it.jsonObject }.map { LifeThemeGroup(it.string("id"), it.string("name")) }
            val themes = root.getValue("themes").jsonArray.map { it.jsonObject }.map { t ->
                LifeTheme(
                    id = t.string("id"),
                    group = t.string("group"),
                    name = t.string("name"),
                    description = t.string("description"),
                    synonyms = t.strings("synonyms"),
                    passages = t.strings("refs").mapNotNull(VerseRange::parse),
                    naveTopics = t.strings("nave"),
                )
            }
            return LifeThemeCatalog(groups, themes)
        }

        /** The catalog keeps each theme's search words as one string, separated by commas. */
        fun splitSynonyms(line: String): List<String> = line.split(',').map { it.trim() }.filter { it.isNotEmpty() }
    }
}

/**
 * How the Topics directory matches what someone types against names and search words — `TopicSearch`
 * in `LifeThemes.swift`, with the same folding and the same scores, so a word finds the same topics on
 * both platforms.
 */
object TopicSearch {
    /**
     * Lowercased, accents and width folded, apostrophes dropped ("God’s" → "gods"), runs of anything
     * that isn't a letter or digit made one space.
     */
    fun normalize(text: String): String {
        // Accents are the combining marks U+0300–036F; Japanese voicing marks are not accents. Composed
        // again afterwards, so a Korean syllable or a voiced kana is one letter again.
        val stripped = Normalizer.normalize(text, Normalizer.Form.NFKD).filter { it.code !in 0x0300..0x036F }
        val folded = Normalizer.normalize(stripped, Normalizer.Form.NFC)
            .lowercase()
            .replace("’", "").replace("'", "")
        val out = StringBuilder()
        var pendingSpace = false
        for (c in folded) {
            if (c.isLetterOrDigit()) {
                if (pendingSpace && out.isNotEmpty()) out.append(' ')
                pendingSpace = false
                out.append(c)
            } else {
                pendingSpace = true
            }
        }
        return out.toString()
    }

    /** Two characters in Chinese, Japanese and Korean, where two make a word; three elsewhere. */
    fun minimumLength(text: String): Int = if (text.any(::isCjk)) 2 else 3

    private fun isCjk(c: Char): Boolean = c.code in 0x3040..0x30FF || c.code in 0x3400..0x9FFF || c.code in 0xAC00..0xD7AF

    /**
     * How well a normalized [term] answers a normalized [query]; 0 for not at all. 100 the term is the
     * query; 80 the query begins it; 70 it is a whole word or phrase inside the query; 60 the query is
     * or begins a word of a longer term. Chinese, Japanese and Korean have no spaces between words, so
     * there containment either way counts.
     */
    fun score(term: String, query: String): Int {
        if (term.isEmpty() || query.isEmpty()) return 0
        if (term == query) return 100
        if (term.startsWith(query)) return 80
        if (term.any(::isCjk) || query.any(::isCjk)) {
            if (term.length >= 2 && query.contains(term)) return 70
            if (query.length >= 2 && term.contains(query)) return 60
            return 0
        }
        if (term.length >= 3 && " $query ".contains(" $term ")) return 70
        if (term.split(' ').any { it.startsWith(query) }) return 60
        return 0
    }
}

File: ScriptureAloneCore/Sources/ScriptureAloneCore/Topics/LifeThemes.swift
import Foundation

/// A heading in the Topics directory — "Worry & Hard Feelings", "Who God Is".
public struct LifeThemeGroup: Hashable, Sendable, Identifiable, Codable {
    public let id: String
    /// English; the catalog key for `localizedName`.
    public let name: String

    /// The name in the language the app is running in.
    public var localizedName: String { LifeThemeCatalog.localized(name) }
}

/// Something a reader may be going through — anxiety, grief, a new job — and passages that speak
/// to it, curated in `Data/topics/life-themes.json` and built by `Tools/build_topics.py`.
///
/// Passages are references only (KJV keys, the numbering everything in the app is stored in): the
/// words always come from the translation being read, so a theme works the same in a bundled,
/// imported, packaged or online translation, in any language.
public struct LifeTheme: Hashable, Sendable, Identifiable {
    /// Stable: never renamed, so anything that remembers a theme keeps finding it.
    public let id: String
    public let group: String
    /// English; the catalog key for `localizedName`.
    public let name: String
    /// English; the catalog key for `localizedDescription`.
    public let description: String
    /// English search words — "worried", "nervous", "panic".
    public let synonyms: [String]
    /// The passages, in the order they were chosen (the strongest first).
    public let passages: [VerseRange]
    /// Nave's Topical Bible topics that cover the same ground, by name (`TopicalIndex`).
    public let naveTopics: [String]

    public init(id: String, group: String, name: String, description: String, synonyms: [String],
                passages: [VerseRange], naveTopics: [String] = []) {
        self.id = id
        self.group = group
        self.name = name
        self.description = description
        self.synonyms = synonyms
        self.passages = passages
        self.naveTopics = naveTopics
    }

    public var localizedName: String { LifeThemeCatalog.localized(name) }
    public var localizedDescription: String { LifeThemeCatalog.localized(description) }

    /// Search words in the language the app is running in. The catalog holds each theme's as one
    /// comma-separated string, keyed by the English list.
    public var localizedSynonyms: [String] {
        let english = synonyms.joined(separator: ", ")
        let localized = LifeThemeCatalog.localized(english)
        guard localized != english else { return [] }
        return localized.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
    }
}

/// The curated life themes, from `LifeThemes.json` in this package's resources.
public struct LifeThemeCatalog: Sendable {
    public let groups: [LifeThemeGroup]
    /// In directory order: group by group, as the source file lists them.
    public let themes: [LifeTheme]

    /// The bundled catalog. Empty only if the resource is missing, which the tests forbid.
    public static let shared: LifeThemeCatalog = {
        guard let url = Bundle.module.url(forResource: "LifeThemes", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let catalog = try? LifeThemeCatalog(data: data) else { return LifeThemeCatalog(groups: [], themes: []) }
        return catalog
    }()

    public init(groups: [LifeThemeGroup], themes: [LifeTheme]) {
        self.groups = groups
        self.themes = themes
    }

    private struct File: Decodable {
        struct Theme: Decodable {
            let id, group, name, description: String
            let synonyms: [String]
            let refs: [String]
            let nave: [String]?
        }
        let version: Int
        let groups: [LifeThemeGroup]
        let themes: [Theme]
    }

    public init(data: Data) throws {
        let file = try JSONDecoder().decode(File.self, from: data)
        groups = file.groups
        themes = file.themes.map { theme in
            LifeTheme(id: theme.id, group: theme.group, name: theme.name, description: theme.description,
                      synonyms: theme.synonyms, passages: theme.refs.compactMap(VerseRange.init(storageString:)),
                      naveTopics: theme.nave ?? [])
        }
    }

    public func theme(id: String) -> LifeTheme? { themes.first { $0.id == id } }
    public func themes(in group: LifeThemeGroup) -> [LifeTheme] { themes.filter { $0.group == group.id } }

    /// Themes a search speaks to, best first: "anxious", "worried sick", "grief", "不安".
    ///
    /// Matches a theme's name and search words, in English and in the app's language. A whole
    /// word is the strongest match; then a word being typed ("anxi"); then a phrase inside what was
    /// typed ("I feel so alone" finds Loneliness). Nothing below two characters, or three in an
    /// alphabet where two letters are rarely a word.
    public func search(_ query: String, limit: Int = 3) -> [LifeTheme] {
        let needle = TopicSearch.normalize(query)
        guard needle.count >= TopicSearch.minimumLength(needle) else { return [] }
        let scored = themes.enumerated().compactMap { index, theme -> (score: Int, index: Int, theme: LifeTheme)? in
            let terms = [theme.name, theme.localizedName] + theme.synonyms + theme.localizedSynonyms
            let score = terms.map { TopicSearch.score(TopicSearch.normalize($0), against: needle) }.max() ?? 0
            return score > 0 ? (score, index, theme) : nil
        }
        return scored.sorted { ($0.score, -$0.index) > ($1.score, -$1.index) }.prefix(limit).map(\.theme)
    }

    /// A catalog string in the app's language, or the English it is keyed by.
    static func localized(_ english: String) -> String {
        Bundle.module.localizedString(forKey: english, value: english, table: nil)
    }
}

/// How the Topics directory matches what someone types against names and search words. Shared by
/// the life themes and Nave's index so the two rank alike.
public enum TopicSearch {
    /// Lowercased, accents and width folded, apostrophes dropped ("God’s" → "gods"), runs of
    /// anything that isn't a letter or digit made one space.
    public static func normalize(_ text: String) -> String {
        let folded = text.folding(options: [.caseInsensitive, .diacriticInsensitive, .widthInsensitive], locale: nil)
            .replacingOccurrences(of: "’", with: "").replacingOccurrences(of: "'", with: "")
        var result = ""
        var pendingSpace = false
        for character in folded {
            if character.isLetter || character.isNumber {
                if pendingSpace, !result.isEmpty { result.append(" ") }
                pendingSpace = false
                result.append(character)
            } else {
                pendingSpace = true
            }
        }
        return result
    }

    /// Two characters in Chinese, Japanese and Korean, where two make a word; three elsewhere.
    public static func minimumLength(_ text: String) -> Int {
        text.unicodeScalars.contains(where: isCJK) ? 2 : 3
    }

    static func isCJK(_ scalar: Unicode.Scalar) -> Bool {
        (0x3040...0x30FF).contains(scalar.value) || (0x3400...0x9FFF).contains(scalar.value)
            || (0xAC00...0xD7AF).contains(scalar.value)
    }

    /// How well a normalized term answers a normalized query; 0 for not at all.
    ///
    /// - 100: the term is the query.
    /// - 80: the query begins the term — a word still being typed.
    /// - 70: the term is a whole word or phrase inside the query ("feeling anxious" has "anxious").
    /// - 60: the query is a whole word of a longer term, or begins one ("health" in "mental health").
    /// Chinese, Japanese and Korean have no spaces between words, so there containment either way
    /// counts as 70.
    public static func score(_ term: String, against query: String) -> Int {
        guard !term.isEmpty, !query.isEmpty else { return 0 }
        if term == query { return 100 }
        if term.hasPrefix(query) { return 80 }
        if term.unicodeScalars.contains(where: isCJK) || query.unicodeScalars.contains(where: isCJK) {
            if term.count >= 2, query.contains(term) { return 70 }
            if query.count >= 2, term.contains(query) { return 60 }
            return 0
        }
        if term.count >= 3, " \(query) ".contains(" \(term) ") { return 70 }
        if term.split(separator: " ").contains(where: { $0.hasPrefix(query) }) { return 60 }
        return 0
    }
}
